### ISSUE TITLE
Possible issues with parsing tokens that contain commas

### ISSUE THREAD

**Issue by mishazharov** [g1eny0ung/parse_link_header#16]
In [RFC 8288](https://datatracker.ietf.org/doc/html/rfc8288#section-3) we see that `token`s can be used in the link header. `token` is defined in [RFC 7230](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6), and can be a quoted entry which contains commas. As a result `"one,two"` is a valid token.

Therefore, I would expect
```
<https://api.github.com/repositories/41986369/contributors?page=2>; "one,two"="next", <https://api.github.com/repositories/41986369/contributors?page=14>; rel="last"
```

to be parsed correctly, however it panics instead. I'm not sure how often this occurs in the wild, and I don't currently require the correct behaviour, but wanted to add an issue in case someone else needs higher compliance with the RFCs

**Comment by g1eny0ung** [comment]
> however it panics instead.

Yep. Because the code split multiple links with `,`. This will cause unexpected behavior.

>   Most HTTP header field values are defined using common syntax
   components (token, quoted-string, and comment) separated by
   whitespace or specific delimiting characters.  Delimiters are chosen
   from the set of US-ASCII visual characters not allowed in a token
   (DQUOTE and "(),/:;<=>?@[\]{}").

But I found that [rfc7230#section-3.2.6](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6) says that the token can't contain some delimiters like `,`. 🤔

**Comment by mishazharov** [comment]
> But I found that [rfc7230#section-3.2.6](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6) says that the token can't contain some delimiters like `,`. :thinking:

You're right, this is true.

> `token` is defined in [RFC 7230](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6), and can be a quoted entry which contains commas. As a result `"one,two"` is a valid token.

What I said earlier is not accurate. What I meant was that the link param is defined as

```
     link-param = token BWS [ "=" BWS ( token / quoted-string ) ]
```
([rfc8288#section-3](https://datatracker.ietf.org/doc/html/rfc8288#section-3))

And the quoted-string can contain a `,` ([rfc7230#section-3.2.6](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6)) which breaks the parsing for the reason you mentioned

**Comment by g1eny0ung** [comment]
> And the quoted-string can contain a `,` ([rfc7230#section-3.2.6](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6)) which breaks the parsing for the reason you mentioned

Yes. The quoted-string can contain a `,` but it is only used as a `value (key=value)` in the definition. In the example you said before, the `"one,two"` is a key.

But even then, the code still does not handle such cases correctly if a link param is `key="foo,bar"`.

**Comment by g1eny0ung** [comment]
@mishazharov I will update the code to solve this problem, but not very quickly, probably within the next two weeks.

**Comment by mishazharov** [comment]
Thank you for taking the time to triage this issue. I believe that the same may apply to `"`, `>`, `<`, and `;`. As well, the quoted string supports escape sequences (`\`). It might be easier to write a parser using https://github.com/rust-bakery/nom to solve this problem.

Also I can try to implement this if you're open to PRs

**Comment by g1eny0ung** [comment]
Sure! Definitely, PRs always welcome.

**Comment by g1eny0ung** [comment]
> It might be easier to write a parser using https://github.com/rust-bakery/nom to solve this problem.

Inspired. I also think so.

**Comment by mishazharov** [comment]
After playing around with the RFC for a while, I decided implement the crate in a different manner, my implementation is here: https://github.com/mishazharov/nom-rfc8288

That took a bit longer than I was expecting though :sweat_smile:
